test/cloudtest/test_k8s_node_recovery.py

```diff
@@ -0,0 +1,387 @@
+    ]
+    # string needs the same indentation as testdrive script below
+    drop_cluster_statement_sql = "\n            ".join(drop_cluster_statements)
+
+    create_cluster_statements = [
+        f"> CREATE CLUSTER {cluster.name} REPLICAS ({cluster.create_replica_definitions_sql()});"
+        for cluster in all_clusters
+    ]
+    # string needs the same indentation as testdrive script below
+    create_cluster_statement_sql = "\n".join(create_cluster_statements)
+
+    mz.testdrive.run(
+        input=dedent(
+            """
+            > DROP MATERIALIZED VIEW IF EXISTS mv;
+            > DROP SOURCE IF EXISTS source;
+
+            """
+            + dedent(drop_cluster_statement_sql)
+            + "\n"
+            + dedent(create_cluster_statement_sql)
+            + "\n"
+            + dedent(
+                f"""
+
+            > SELECT name FROM mz_clusters WHERE name IN ('{compute_cluster.name}', '{storage_cluster.name}');
+            {storage_cluster.name}
+            {compute_cluster.name}
+
+            > CREATE SOURCE source IN CLUSTER {storage_cluster.name}
+              FROM LOAD GENERATOR COUNTER
+              (TICK INTERVAL '500ms');
+
+            > SELECT COUNT(*) FROM (SHOW SOURCES) WHERE name = 'source';
+            1
+
+            > CREATE MATERIALIZED VIEW mv (f1) IN CLUSTER {compute_cluster.name} AS SELECT * FROM source;
+
+            > CREATE DEFAULT INDEX IN CLUSTER {compute_cluster.name} ON mv;
+
+            > SELECT COUNT(*) > 0 from mv;
+            true
+            """
+            )
+        ),
+        no_reset=True,
+    )
+
+
+def validate_state(
+    mz: MaterializeApplication,
+    reached_index: int,
+    must_exceed_reached_index: bool,
+    timeout_in_sec: int,
+    expected_state: str,
+) -> None:
+    comparison_operator = ">" if must_exceed_reached_index else ">="
+    print(f"Expect '{expected_state}' within timeout of {timeout_in_sec}s")
+
+    testdrive_timeout_in_sec = 10
+
+    validation_succeeded = False
+    last_error_message = None
+
+    start_time = time.time()
+
+    # re-run testdrive to make sure it connects to the most recent envd
+    for run in range(0, int(timeout_in_sec / testdrive_timeout_in_sec)):
+        try:
+            mz.testdrive.run(
+                input=dedent(
+                    f"""
+                    > SELECT COUNT(*) {comparison_operator} {reached_index} FROM source; -- validate source
+                    true
+    
+                    > SELECT COUNT(*) {comparison_operator} {reached_index} FROM mv; -- validate mv
+                    true
+                    """
+                ),
+                default_timeout=f"{testdrive_timeout_in_sec}s",
+                no_reset=True,
+            )
+            validation_succeeded = True
+            break
+        except Exception as e:
+            # arbitrary error can occur if envd is not yet ready after restart
+            print("Error occurred, retrying.")
```

Comment:
I think this causes too much output to happen even for expected errors, causing the CI output to be very cluttered. Consider printing the full error only after the retries have been exhausted.

Reply:
Yes, it is true that there is too much output. Though, I think it is not this statement causing the mess but testdrive itself, which I will need to instruct not to print on failures.